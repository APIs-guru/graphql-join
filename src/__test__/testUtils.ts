declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionArgs,
  ExecutionResult,

  graphql,
  parse,
  validate,
  execute,
  print,
  buildSchema,
  printSchema,
  formatError,
  specifiedRules,
} from 'graphql';
import * as _ from 'lodash';

import { GraphQLJoinSchema, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, args: object, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    let result = `${schemaName}::${parentType.name}::${fieldName}`;
    if (Object.keys(args).length !== 0) {
      result += `;args=${JSON.stringify(args)}`
    }
    return result;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode, variableValues?: object) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    // FIXME: Should use executeQuery but blocked by
    // https://github.com/facebook/jest/issues/3917
    return graphql({
      schema,
      source: new Source(query, 'Send to ' + schemaName),
      variableValues,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string | { idl: string, prefix?: string };
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinIDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (schemaSource, name) => {
    if (typeof schemaSource === 'string') {
      schemaSource = { idl: schemaSource };
    }
    const { prefix, idl } = schemaSource;
    const schema = buildSchema(new Source(idl, name));
    stubSchema(schema);
    return { schema , prefix};
  });

  const joinSchema = new GraphQLJoinSchema(
    new Source(joinIDL, 'Join SDL'),
    remoteSchemas
  );
  const schema = joinSchema.schema;

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string | { query: string, variableValues: object },
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (queryAST) => {
          const query = print(queryAST);
          expect(query).toMatchSnapshot();
          return Promise.resolve(results[name])
        };
      }
      return makeProxy(name, schema)
    });

    const queryObj = typeof query === 'string' ?
      { query: query, variableValues: {} } : query;
    const result = await executeQuery({
      schema,
      document: parse(new Source(queryObj.query, 'ClientQuery')),
      variableValues: queryObj.variableValues,
      contextValue: new ProxyContext(joinSchema, proxyFns),
    });

    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function executeQuery(args: ExecutionArgs) {
  const validationErrors = validate(args.schema, args.document, specifiedRules);
  expect(validationErrors.join('\n')).toBe('');
  return execute(args);
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

