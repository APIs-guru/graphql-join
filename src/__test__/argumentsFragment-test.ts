import { testJoin } from './testUtils';

describe('arguments fragment tests', () => {
  test('get argument from fragment', async () => {
    const execute = testJoin(
      {
        test1: `
          type TestType {
            foo: String
          }
          type Query1 {
            testObj: TestType
          }
          schema { query: Query1 }
        `,
        test2: `
          type Query {
            bar(fooArg: String): String
          }
        `,
      }, `
       schema { query: Query1 }
       extend type TestType {
         bar: String @resolveWith(query: "bar", argumentsFragment: "FooArg")
       }
       query bar($foo: String) @send(to: "test2") {
         bar(fooArg:$foo)
       }
       fragment FooArg on TestType {
         foo @export(as: "foo")
       }
      `
    );
    await execute('{ testObj { bar } }');
  });
});
