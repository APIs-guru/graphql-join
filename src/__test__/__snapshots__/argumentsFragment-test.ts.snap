// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arguments fragment tests conflicting fields in argumentsFragment and client selection 1`] = `
type Foo {
  bar(barArg: String): String
  baz: String
}

type Query {
  foo: Foo
}
`;

exports[`arguments fragment tests conflicting fields in argumentsFragment and client selection 2`] = `
{
  foo {
    bar(barArg: "ClientValue")
    ___f_BazArgs_bar: bar(barArg: "FragmentValue")
  }
}
`;

exports[`arguments fragment tests conflicting fields in argumentsFragment and client selection 3`] = `
{
  baz(barValue: "test1::Foo::bar;args={\\"barArg\\":\\"FragmentValue\\"}")
}
`;

exports[`arguments fragment tests conflicting fields in argumentsFragment and client selection 4`] = `
[
  "\\n      {\\n        foo {\\n          bar(barArg: \\"ClientValue\\")\\n          baz\\n        }\\n      }\\n    ",
  {
    "data": {
      "foo": {
        "bar": "test1::Foo::bar;args={\\"barArg\\":\\"ClientValue\\"}",
        "baz": "test2::Query::baz;args={\\"barValue\\":\\"test1::Foo::bar;args={\\\\\\"barArg\\\\\\":\\\\\\"FragmentValue\\\\\\"}\\"}"
      }
    },
    "errors": []
  }
]
`;

exports[`arguments fragment tests get argument from fragment 1`] = `
schema {
  query: Query1
}

type Query1 {
  testObj: TestType
}

type TestType {
  foo: String
  bar: String
}
`;

exports[`arguments fragment tests get argument from fragment 2`] = `
{
  testObj {
    ___f_FooArg_foo: foo
  }
}
`;

exports[`arguments fragment tests get argument from fragment 3`] = `
{
  bar(fooArg: "test1::TestType::foo")
}
`;

exports[`arguments fragment tests get argument from fragment 4`] = `
[
  "{ testObj { bar } }",
  {
    "data": {
      "testObj": {
        "bar": "test2::Query::bar;args={\\"fooArg\\":\\"test1::TestType::foo\\"}"
      }
    },
    "errors": []
  }
]
`;
